source/h1_encoder.c

```diff
@@ -319,6 +344,309 @@ int aws_h1_encoder_start_message(
+            }
+
+            ENCODER_LOG(
+                TRACE,
+                encoder,
+                "No body data written, concluding this message. "
+                "Will try to write body data again in the next message.");
+            return AWS_OP_SUCCESS;
+        }
+    }
+    return AWS_OP_SUCCESS;
+}
+
+bool write_crlf(struct aws_byte_buf *dst) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    char crlf[CRLF_SIZE] = {'\r', '\n'};
+    struct aws_byte_cursor crlf_cursor = aws_byte_cursor_from_array(&crlf, AWS_ARRAY_SIZE(crlf));
+    return aws_byte_buf_write_from_whole_cursor(dst, crlf_cursor);
+}
+
+bool write_chunk_size(struct aws_byte_buf *dst, struct aws_input_stream *chunk) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(chunk);
+    AWS_PRECONDITION((dst->capacity - dst->len) >= MAX_ASCII_HEX_CHUNK_STR_SIZE);
+    int64_t chunk_size = 0;
+    if (AWS_OP_SUCCESS != aws_input_stream_get_length(chunk, &chunk_size)) {
+        return false;
+    }
+    char ascii_hex_chunk_size_str[MAX_ASCII_HEX_CHUNK_STR_SIZE];
+    AWS_ZERO_ARRAY(ascii_hex_chunk_size_str);
+    snprintf(ascii_hex_chunk_size_str, sizeof(ascii_hex_chunk_size_str), "%zX", chunk_size);
+    return aws_byte_buf_write_from_whole_cursor(dst, aws_byte_cursor_from_c_str(ascii_hex_chunk_size_str));
+}
+
+bool write_chunk_extension(struct aws_byte_buf *dst, struct aws_http1_chunk_extension *chunk_extension) {
+    AWS_PRECONDITION(chunk_extension);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    bool wrote_all = true;
+    wrote_all &= aws_byte_buf_write_u8(dst, ';');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(dst, chunk_extension->key);
+    wrote_all &= aws_byte_buf_write_u8(dst, '=');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(dst, chunk_extension->value);
+    return wrote_all;
+}
+
+static bool s_fill_byte_buffer(struct aws_byte_buf *dst, struct aws_byte_cursor *src) {
+    size_t write_size = aws_min_size(dst->capacity - dst->len, src->len);
+    struct aws_byte_cursor sub_cursor;
+    sub_cursor.len = write_size;
+    sub_cursor.ptr = src->ptr;
+    if (AWS_UNLIKELY(AWS_OP_SUCCESS != aws_byte_buf_append(dst, &sub_cursor))) {
+        return false;
+    }
+    aws_byte_cursor_advance(src, write_size);
+    return 0 == src->len;
+}
+
+static bool s_chunk_line_state(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(encoder->message);
+    AWS_PRECONDITION(encoder->message->body_chunk);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(aws_byte_buf_is_valid(&encoder->message->body_chunk->chunk_line));
+    bool completely_written = s_fill_byte_buffer(dst, &encoder->message->body_chunk->chunk_line_cursor);
+    if (completely_written) {
+        aws_byte_buf_clean_up(&encoder->message->body_chunk->chunk_line);
```

Comment:
Not 100% sure: Are chunk_line, and anything else chunk-related, getting cleaned upeven if the connection comes to a halt and stuff never gets sent? We also need to be sure completion callbacks fire so users know that it's safe to clean up the aws_input_streams. You could add a test that queues up a bunch of chunks and then just closes the connection.

Reply:
Yep. I have added a test. Since we added internal allocation, there's several paths which now need to be cleaned up on error cases. I will have remedied these in the next commit update.